# Kovaryans ve Korelasyon İlişkisi
Bu konuda, Korelasyon'un (Konu 82) Kovaryans'tan (Varyans-Kovaryans 
Matrisi konusu) nasıl **türetildiğini**, somut sayılarla göreceğiz.

## Formülü Hatırlayalım
$$r = \frac{Cov(X,Y)}{s_X \cdot s_Y}$$

## Neden Bu Bölme İşlemi Sonucu -1 ile +1 Arasına Sıkıştırıyor?
Bu, aslında biraz ileri bir matematiksel gerçek (Cauchy-Schwarz 
Eşitsizliği'nden gelir) ama sezgisel olarak şöyle düşünülebilir: 
Kovaryansın büyüklüğü, iki değişkenin **kendi varyanslarının çarpımını** 
asla geçemez — yani $Cov(X,Y)^2 \le Var(X) \cdot Var(Y)$ her zaman 
doğrudur. Kovaryansı, $s_X \cdot s_Y$'ye (yani $\sqrt{Var(X) \cdot Var(Y)}$'ye) 
bölünce, bu doğal üst sınır sayesinde sonuç hep -1 ile +1 arasında kalır.

## Somut Hesaplama
Varyans-Kovaryans Matrisi konusundaki (Reklam Harcaması - Satış Adedi) 
örneğimizi hatırlayalım: $Cov(Reklam, Satış) = 6894.21$

Bu tek başına "büyük mü küçük mü" bilemiyorduk. Şimdi korelasyona 
çevirelim:
```python
import numpy as np

kovaryans = 6894.21
std_reklam = np.sqrt(8100.12)  # Varyansın karekökü = std sapma (Konu 12)
std_satis = np.sqrt(8014.03)

korelasyon = kovaryans / (std_reklam * std_satis)
print(korelasyon)
```

## Python'da Doğrudan Hesaplama
Elle hesaplamak yerine, pandas/numpy zaten hazır fonksiyonlar sunuyor:
```python
df[['Reklam_Harcamasi','Satis_Adedi']].corr()  # pandas ile
np.corrcoef(reklam, satis)                      # numpy ile
```

## Kovaryans Matrisi ile Korelasyon Matrisi Arasındaki Paralellik
Tıpkı Varyans-Kovaryans Matrisi'nde olduğu gibi, birden fazla değişken 
için bir **Korelasyon Matrisi** de oluşturulabilir — köşegen hep 1'dir 
(bir değişkenin kendisiyle korelasyonu her zaman mükemmel, r=1), köşegen 
dışı elemanlar ise ikili korelasyonları gösterir.

In [ ]:
import numpy as np
import pingouin as pg
ders_calisma = [0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5, 5.5, 6, 6.5, 7, 7.5, 8, 8.5, 9, 9.5, 10]
alinan_not = [5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95, 100]
korelasyon = pg.corr(x=ders_calisma, y=alinan_not)
print(korelasyon)

          n    r        CI95          p_val        BF10  power
pearson  20  1.0  [1.0, 1.0]  1.245968e-139  2.351e+122    1.0


In [6]:
# ders çalışma ve alınan not arasındaki kovaryans matrisini önce otomatik hesaplayıp ardından elle hesaplayacağız.
covar = np.cov(m=ders_calisma, y=alinan_not)
print(covar)

[[  8.75  87.5 ]
 [ 87.5  875.  ]]


Ders çalışma süresi ile alınan not arasındaki kovaryans 87.5 çıktı — bu 
ham sayı tek başına "ilişkinin ne kadar güçlü" olduğunu anlamamızı 
sağlamıyor (Varyans-Kovaryans Matrisi konusunda tartıştığımız gibi). Bu 
yüzden standartlaştırıp korelasyona çevirmemiz gerekiyor.

In [11]:
# manuel korelasyon hesabı
x_y_kovaryans = covar[0][1]
std_x = np.std(ders_calisma, ddof=1)
std_y = np.std(alinan_not, ddof=1)
correlation = x_y_kovaryans / (std_x * std_y)
print(f'Ders çalışma süresi ve alınan not kovaryans değeri: {x_y_kovaryans}')
print(f'Ders çalışma sürelerinin standart sapması: {std_x}')
print(f'Alınan notların standart sapması: {std_y}')
print(f'Korelasyon katsayısı: {correlation}')

Ders çalışma süresi ve alınan not kovaryans değeri: 87.5
Ders çalışma sürelerinin standart sapması: 2.958039891549808
Alınan notların standart sapması: 29.58039891549808
Korelasyon katsayısı: 1.0


Ders çalışma süresini 30 dk arttırdığımızda alınan puanın 5 puan artacağı şekilde bir yapay veri seti kurguladık bu yüzden tam doğrusallık oluştu ve korelasyon katsayısı maksimum çıktı ancak gerçek dünya verileri bu kadar kusursuz olmayabilir işin içine gürültü girecektir.